src/geometry/detail/VGNavCollection.hh

```diff
@@ -0,0 +1,204 @@
+    //! True if the collection is assigned/valiid
+    explicit operator bool() const { return static_cast<bool>(ptr); }
+};
+#endif
+
+//---------------------------------------------------------------------------//
+// DEVICE MEMSPACE
+//---------------------------------------------------------------------------//
+/*!
+ * Delete a VecGeom pool.
+ *
+ * Due to VecGeom macros, the definition of this function can only be compiled
+ * from a .cc file.
+ */
+struct NavStatePoolDeleter
+{
+    using arg_type = vecgeom::cxx::NavStatePool*;
+    void operator()(arg_type) const;
+};
+
+#ifndef __CUDA_ARCH__
+//---------------------------------------------------------------------------//
+/*!
+ * Manage a pool of device-side geometry states.
+ *
+ * Construction and destruction of the NavStatePool has to be in a host
+ * compilation unit due to VecGeom macro magic. We hide this class to keep
+ * NavStatePool and smart pointer usage from the NVCC device compiler.
+ */
+template<>
+struct VGNavCollection<Ownership::value, MemSpace::device>
+{
+    using UPNavStatePool
+        = std::unique_ptr<vecgeom::cxx::NavStatePool, NavStatePoolDeleter>;
+
+    UPNavStatePool pool;
+    void*          ptr  = nullptr;
+    size_type      size = 0;
+
+    // Resize based on geometry params and state size
+    void resize(int max_depth, size_type size);
+    //! True if the collection is assigned/valid
+    explicit CELER_FUNCTION operator bool() const { return ptr; }
+};
+#endif
+
+//---------------------------------------------------------------------------//
+/*!
+ * Reference on-device memory owned by VGNavCollection<value, device>.
+ *
+ * The NavStatePool underpinning the storage returns a void pointer that must
+ * be manually manipulated to get a single state pointer.
+ */
+template<>
+struct VGNavCollection<Ownership::reference, MemSpace::device>
+{
+    using NavState = vecgeom::cuda::NavigationState;
+
+    void* ptr = nullptr;
+    size_type size = 0;
+
+    // Assign from device value
+    void operator=(VGNavCollection<Ownership::value, MemSpace::device>& other);
+    // Get the navigation state for the given thread
+    inline CELER_FUNCTION NavState& at(int max_depth, ThreadId thread) const;
+    //! True if the collection is assigned/valiid
+    explicit CELER_FUNCTION operator bool() const { return ptr; }
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the navigation state at the given thread.
+ */
+CELER_FUNCTION auto VGNavCollection<Ownership::reference, MemSpace::device>::at(
+    int max_depth, ThreadId thread) const -> NavState&
+{
+    CELER_EXPECT(ptr);
+    CELER_EXPECT(max_depth > 0);
+    CELER_EXPECT(thread < size);
+#ifdef __NVCC__
+    // This code only compiles when run through CUDA so it must be escaped.
+    char* result = reinterpret_cast<char*>(this->ptr);
+    result += NavState::SizeOfInstanceAlignAware(max_depth) * thread.get();
+    return *reinterpret_cast<NavState*>(result);
```

Comment:
As a side note this is the implementation of `NavigationPool::operator[]`.

At the moment it is not `practical` to use it as `vecgeom::cuda::NavStatePool` is missing a constructor (and destructor) to build one from already allocated memory.  

Side side note, if this class had the data members:
```
  int fCapacity; // the number of states in the pool
  int fDepth;    // depth of the navigation objects to cover
  char *fBuffer; // the memory buffer in which we place states
```
the following C++ (fragile) code would work:
```
return reinterpret_cast<vecgeom::NavStatePool>(this)->operator[thread.get()];
```

> // This code only compiles when run through CUDA so it must be escaped.

I am not sure why that is the case, the same code is in host-device function in NavStatePool ...

Reply:
> ```c++
> return reinterpret_cast<vecgeom::NavStatePool>(this)->operator[thread.get()];
> ```
😱